src/v/cluster/producer_state.h

```diff
@@ -0,0 +1,231 @@
+    ///   is no longer in the list of producers tracked.
+    /// - Re-registration helps the manager track LRU-ness, re-registration
+    ///   effectively puts this producer at the end of the queue when
+    ///   considering candidates for eviction.
+    template<AcceptsUnits AsyncFunc>
+    auto run_func(AsyncFunc&& func) {
+        return ss::with_gate(
+          _gate, [this, func = std::forward<AsyncFunc>(func)]() mutable {
+              return _op_lock.get_units().then(
+                [this, f = std::forward<AsyncFunc>(func)](auto units) {
+                    unlink_self();
+                    _ops_in_progress++;
+                    return ss::futurize_invoke(f, std::move(units))
+                      .then_wrapped([this](auto result) {
+                          _ops_in_progress--;
+                          link_self();
+                          return result;
+                      });
+                });
+          });
+    }
+
+    ss::future<> shutdown_input();
+    ss::future<> evict();
+    bool is_evicted() const { return _evicted; }
+
+    requests::request_result_t
+    enqueue_request(const model::batch_identity&, model::term_id current_term);
+    void update(const model::batch_identity&, kafka::offset);
+
+    std::optional<seq_t> last_sequence_number() const;
+
+    producer_state_snapshot snapshot(kafka::offset log_start_offset) const;
+
+private:
+    // Register/deregister with manager.
+    void register_self();
+    void deregister_self();
+    // Utilities to temporarily link and unlink from manager
+    // without modifying the producer count.
+    void link_self();
+    void unlink_self();
+
+    std::chrono::milliseconds ms_since_last_update() const {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(
+          ss::lowres_system_clock::now() - _last_updated_ts);
+    }
+
+    void tickle() { _last_updated_ts = ss::lowres_system_clock::now(); }
+
+    model::producer_identity _id;
+    raft::group_id _group;
+    // serializes all the operations on this producer
+    mutex _op_lock;
+    std::reference_wrapper<producer_state_manager> _parent;
+
+    requests _requests;
+    // Tracks the last time an operation is run with this producer.
+    // Used to evict stale producers.
+    ss::lowres_system_clock::time_point _last_updated_ts;
+    intrusive_list_hook _hook;
+    ss::gate _gate;
+    // function hook called on eviction
+    bool _evicted = false;
+    size_t _ops_in_progress = 0;
+    ss::noncopyable_function<void()> _post_eviction_hook;
+    friend class producer_state_manager;
+    friend struct ::test_fixture;
+};
+
+struct producer_state_snapshot {
+    struct finished_request {
+        seq_t _first_sequence;
+        seq_t _last_sequence;
+        kafka::offset _last_offset;
+    };
+
+    model::producer_identity _id;
+    raft::group_id _group;
+    fragmented_vector<finished_request> _finished_requests;
```

Comment:
do we expect the finished_requests vector to be large ?

Reply:
no, its bounded to 5.

Reply:
than i think fragmented_vector is not a good fit here

Reply:
are you suggesting that there is wastage because the actual contents are much smaller than a single allocated fragment? Just confirming, so vector + reserve?

Reply:
exactly